src/boot/efi/log.c

```diff
@@ -7,16 +7,48 @@
 
 static unsigned log_count = 0;
 
+#define DEBUGCON_PREFIX u"systemd: "
+#define DEBUGCON_ACK 0xE9
+#define DEBUGCON_PORT 0x402
+
 void freeze(void) {
         for (;;)
                 BS->Stall(60 * 1000 * 1000);
 }
 
+#if defined(__i386__) || defined(__x86_64__)
+static bool log_has_debugcon(void) {
+        static bool init;
+        static bool present;
+
+        if (!init) {
+                present = (inb(DEBUGCON_PORT) == DEBUGCON_ACK);
+                init = true;
+        }
+        return present;
+}
+
+static void log_debugcon(const char16_t *msg) {
+        size_t i;
+
+        if (!log_has_debugcon())
+                return;
+
+        for (i = 0; msg[i]; i++)
+                outb(DEBUGCON_PORT, msg[i]);
```

Comment:
hmm, this is a bit sloppy, since you just drop the upper 8 byte as is. maybe do:

```c
outb(DEBUGCON_PORT, (uint8_t) msg[i] < 0x7FU ? msg[i] : '?');
```

or so?